### ISSUE TITLE
Requirements for phase 3

### ISSUE THREAD

**Issue by tlively** [WebAssembly/threads#163]
This repo has not had a lot of activity lately, so I'm wondering if this proposal is ready to move to phase 3.

The phase 3 entry requirements are:

 1. Test suite has been updated to cover the feature in its forked repo.

 2. The test suite should run against some implementation, though it need not be the reference interpreter.

We certainly have a working implementation, and it looks like a test suite has been added. Is there anything else we previously decided to wait on before moving to phase 3?

**Comment by lars-t-hansen** [comment]
I thought the process required the spec text but it does not, so we don't have to wait on the memory model for phase 3, for one thing.

(FF, like Chrome, has shipped this feature.)

**Comment by binji** [comment]
Good point, I had assumed we were waiting on some additional spec work to advance, but this looks like it can move to phase 3. The testsuite is on concern, however -- we have surface-level tests of the new instructions, but we'll really want something more comprehensive. Maybe it's sufficient for phase 3 if we can move some of Chrome's and FF's JS tests into this repo.

**Comment by Constellation** [comment]
Note that JSC implemented this behind the flag.
https://trac.webkit.org/changeset/270208/webkit

**Comment by dtig** [comment]
Looks like this discussion has stalled, was this discussed in a meeting with additional follow up? If not, I'm wondering if in addition to the existing spec tests, would [WPT](https://web-platform-tests.org/) be a good place to start to make sure that everything works as it should with workers? 

My concern with JS Shell tests is that while these will work, they are not reflective of real world cases i.e. at least in d8, the worker API is somewhat barebones and depending on test cases it may be possible to get different results when running in the shell vs. the browser, and while we can fix the API, that seems to me to be the wrong thing to fix in this case. I'm probably missing quite a bit of context w.r.t. testing, but I'm wondering what we can do in the near term to hit phase 3 requirements. Concretely I'd like to propose that we decouple comprehensive testing and figuring out which high level test predicate should be used - i.e. use WPT in the near term for comprehensive testing for phase 3, and work on the larger problem of implementing/using the API linked above, or a different strategy for higher level testing separately.

**Comment by tlively** [comment]
> Looks like this discussion has stalled, was this discussed in a meeting with additional follow up?

It was not.

**Comment by rossberg** [comment]
The test suite requirement should be read as tests integrated into the wast-based spec test suite. A JS-specific test suite is of little use to non-JS engines.

As @lars-t-hansen suggested, we should be able to design a suitable extension to the wast format. Would some form of "spawn" command, perhaps paired with a wait command, be sufficient?

**Comment by lars-t-hansen** [comment]
Some test cases for inspiration here:

* https://hg.mozilla.org/mozilla-central/file/tip/js/src/jit-test/tests/wasm/atomicity.js
* https://hg.mozilla.org/mozilla-central/file/tip/js/src/tests/shell/futex-apis.js
* https://hg.mozilla.org/mozilla-central/file/tip/js/src/tests/shell/futex.js

in addition to whatever's in WPT and test262 of course.  Apart from book-keeping it looks like spawning a function in a different thread and a wait or sleep command to give the spawned thread some breathing room to start up and do its thing may be sufficient for something that can claim to at least test all functions in the API.  (The above tests are known to be brittle and have intermittent failures on (very) overloaded test systems; we can live with that, but I don't know if the CG can.)

(If we add something to the wast format, though, the JS generator will need to be able to target the different shells with their idiosyncratic thread mechanisms, or I'll be back on the hook for designing a common system.)

**Comment by rossberg** [comment]
Proper wasm threads would be nice to have. :)

I see that the first of your links already abstracts out spawn and send functions. Do you think it would be feasible for the generator to just produce calls to a small set of such functions and let custom runners define them appropriately?

**Comment by lars-t-hansen** [comment]
> I see that the first of your links already abstracts out spawn and send functions. Do you think it would be feasible for the generator to just produce calls to a small set of such functions and let custom runners define them appropriately?

That's what the agent abstraction in test262 does, and my understanding is that it more or less works OK for them.  ISTR the agent abstraction is fairly complicated because it works for both shells and browsers.  Could we get away with something for shells only?

**Comment by chicoxyzzy** [comment]
According to https://webassembly.org/roadmap/, all major browsers have threads feature implemented already

**Comment by conrad-watt** [comment]
@chicoxyzzy the issue for this phase of standardisation is still the test suite. Thanks for the ping though, we probably want to un-stall this.

@rossberg, you mention above that the test suite requirement should be satisfied by `.wast`-style tests. So I believe as a first stage, we need to bikeshed a syntax for this.

Currently, the `.wast` test suite leaves instantiation pretty implicit. I think to write multithreaded tests, we need the following bare-minimum functionality

- explicitly instantiate a module in a worker thread
- explicitly transfer shareable instantiated exports to a worker thread
- invoke functions across multiple workers simultaneously

Other logic such as number of iterations, blocking/coordination between threads, etc can be expressed in pure Wasm.

So I propose three new test primitives (below). At a high level, worker threads will be indexed by an immediate `n`. We can either explicitly require max `n` to be declared upfront, or assume more will be spawned as necessary. Each worker will be treated as having its own namespace. 

```
instantiate_and_register_on_thread n "instancename" $modulename
```
This primitive takes module `$modulename`, instantiates it in worker thread `n`, and registers the resulting instance in the worker's namespace as `"instancename"`. 

```
transfer_and_register_on_thread n "instancename"
```
Transfers the _shareable_ parts of main-thread-registered instance `"instancename"` to the worker thread `n`, registering it in that thread's namespace with the same name. Currently this would only transfer `shared` memory exports, if any exist.

```
assert_on_threads (n (thread-assert))*

where thread-assert ::=
  assert_return_thread (thread-op) v_res*
| assert_return_one_of_thread (thread-op) (v_res*)*
| assert_trap_thread (thread-op)
| assert_timeout_thread (thread-op)
| ... others?

where thread-op ::=
  invoke_thread "instancename" "funcname" v_args*
| ... others?
```

Simultaneously execute the referenced functions, passing iff all returned results match.

Example `.wast` test:

```wast
;; MP_atomic.wast

(module
  (memory (export "memory") 1 1)
)

(register "test")
(transfer_and_register_on_thread 0 "test")
(transfer_and_register_on_thread 1 "test")

(module $t0
  (memory (import "test" "memory") 1 1)

  (func (export "MP0") (param i32 i32 i32 i32)
    (i32.atomic.store (local.get 0) (local.get 2))
    (i32.atomic.store (local.get 1) (local.get 3))
  )
)

(instantiate_and_register_on_thread 0 "t0" $t0)

(module $t1
  (memory (import "test" "memory") 1 1)

  (func (export "MP1") (param i32 i32) (result i32 i32)
    (i32.atomic.load (local.get 1))
    (i32.atomic.load (local.get 0))
  )
)

(instantiate_and_register_on_thread 1 "t1" $t1)

(assert_on_threads
  (0
    (assert_return_thread
      (invoke_thread "t0" "MP0" (i32.const 0) (i32.const 8) (i32.const 42) (i32.const 1))
    )
   )
  (1
    (assert_return_one_of_thread
      (invoke_thread "t1" "MP1" (i32.const 0) (i32.const 8))
      ((i32.const 0) (i32.const 0))
      ((i32.const 1) (i32.const 42))
    )
   )
)
```

**Comment by rossberg** [comment]
@conrad-watt, that scope makes sense. However, I would suggest a somewhat more orthogonal extension to .wast scripts.

For example, I could imagine a small set of new commands like
```
cmd ::= ... | (thread <name>? <cmd>*) | (wait <name>) | (timeout <name>)
```
The commands inside a thread can access every module or instance from the parent thread, but it will be an error if they try to access an instance that is not transferable -- where an instance is deemed "transferable" if it only has transferable exports, i.e., using instances as proxies for the actual things.

Wouldn't that be enough? I believe everything else can be expressed in this framework (perhaps with a couple more additions). In particular, threads can contain assertions, and by waiting on a thread we can ensure they got executed.

**Comment by conrad-watt** [comment]
I like the idea of waiting on a whole thread instead of on an individual result. My only concern with that design would be the precise semantics of a child thread accessing its parents namespace, especially mapping this onto operations on Web workers so that engines can run the tests.

Could we make it syntactically scoped, so that a child thread gets a deep copy of (only) the shared parts of its parents namespace at the moment it's created? Then it could be modelled as the parent postMessaging all the current in-scope `shared` things created by instantiation to the child at the moment of its creation.

**Comment by rossberg** [comment]
Sounds plausible. In fact, I think it's equivalent: if you copy only the sharable parts, then an attempt to use something else will likewise result in an error (assuming we only make the other names inaccessible, not have them vanish from scope, which would be bad because it could unshadow something earlier).

**Comment by conrad-watt** [comment]
> (assuming we only make the other names inaccessible, not have them vanish from scope, which would be bad because it could unshadow something earlier).

The way I'm thinking of things, the child thread never looks in a parent's namespace, it just starts with a copy and works from that. So the unshared elements do vanish from scope, but they don't unshadow anything. In fact, if a child thread were to shadow/replace a shared thing copied from the parent with an unshared thing, a child-of-child could never access the parent's shared thing.

**Comment by rossberg** [comment]
> In fact, if a child thread were to shadow/replace a shared thing copied from the parent with an unshared thing, a child-of-child could never access the parent's shared thing.

Ah, okay. But with that clarification I think that's also equivalent to what I described. Unless I'm missing something, the only difference would be the error message for a bogus script, like, "instance not shared" vs "inaccessible instance" vs "unbound instance".

**Comment by conrad-watt** [comment]
@rossberg I may have been going too far down the rabbit hole of worrying about a semantics where the parent scope can be updated after the child thread is created in a way that's visible to the child - on reflection it's not clear that this is what you were ever describing :P

So I think your proposed design is the right one - especially with the `one_of` pattern addition. So I guess a good next step would be to implement the parsing of these in the reference interpreter, stubbing out their actual semantic behaviours for now? I can also write a few representative tests.

**Comment by rossberg** [comment]
Actually, on reflection, I was only considering lexically scoped names. But there also is the "registry" used for resolving imports, which is more like a side table. For that I would indeed suggest creating a separate one per thread -- but simply starting empty, not even copying the parent one.

**Comment by conrad-watt** [comment]
> For that I would indeed suggest creating a separate one per thread -- but simply starting empty, not even copying the parent one.

So the mechanism to instantiate two modules in different threads sharing the same memory would be that each thread must first perform `register "instancename" $modulename`, where `$modulename` is defined by a mutual parent and exports the shared memory? Makes sense!

**Comment by rossberg** [comment]
Yes, that's the idea.

**Comment by mattgenious** [comment]
Hi, just trying to figure out if this has stalled.
I apologize if my comment turns out to be noise because of context or knowledge I am missing.

What is currently missing for this proposal to move to phase 3?

I am having a really hard time trying to figure out what is missing and which meeting nodes mentioning threads actually have anything to do with this proposal.

**Comment by tlively** [comment]
@mattgenious, the only thing missing for phase 3 is the test suite. The reason that is tricky is that the current test suite has no concept of threads, so there is some design work necessary to figure out how to extend it properly. That's what the conversation above is about.

Unfortunately (or fortunately?) there hasn't been much motivation to push on this because threads have already been shipped in all the major browsers.

**Comment by conrad-watt** [comment]
FWIW, now that https://github.com/WebAssembly/threads/pull/179 is merged, the only thing missing that would allow us to write  a test suite is the implementation of `wait` and `notify` in the reference interpreter.

**Comment by mattgenious** [comment]
@tlively Thanks for the info, makes a little more sense to me now.
The reason I am interested is because as far as I know the .NET runtime Mono is waiting to implement threads until a spec has been "officially" released and I have a professional interest in that happening because the startup of the mono runtime is blocking the UI thread essentially causing ~2 seconds of unnecessary delay during load.

So I am tracing the path to find where work is missing :-)

@conrad-watt is it correctly understood that the tests need to be implemented in https://github.com/WebAssembly/spec repo 

And that the semantics are already present in the wasm-module-builder.js as kExprAtomicNotify, kExprI32AtomicWait and kExprI64AtomicWait.

I feel like I am making a lot of noise here, is there a better place where I can ask questions without causing trouble?

I joined the W3 work group but haven't seen any activity there.

**Comment by tlively** [comment]
I'll let @conrad-watt fill in the technical details, but you're definitely asking these questions in the right place, @mattgenious.

**Comment by rossberg** [comment]
@conrad-watt, none I would remember. I believe these stubs where left by binji. Feel free to give it a shot.

**Comment by mattgenious** [comment]
Thanks  @conrad-watt, @rossberg and @tlively for kindly including me in the discussion despite my obvious lack of knowledge on the subject.

Alright, I haven't ever written a line of OCaml before and I am also pretty sure I do not have sufficient understanding of the subject matter to venture into that domain.
While I had a course in uni about writing interpreters in F# I think I have to start from scratch to make a dent in anything related to this.

I assume someone will have written the missing code before I learn OCaml and interpreter design, but if not I am willing to put in some effort.

**Comment by conrad-watt** [comment]
We're currently planning to vote on phase 3 at the [in-person meeting](https://github.com/WebAssembly/meetings/blob/main/main/2022/CG-10.md) later this month! Before the vote, I'm planning to add a suite of concurrent tests based on some well-known "litmus" tests for concurrent behaviour [(link)](https://www.cl.cam.ac.uk/~pes20/arm-supplemental/).

**Comment by conrad-watt** [comment]
Threads are now [phase 3](https://github.com/WebAssembly/proposals/pull/152)!
